aQute.libg/src/aQute/lib/concurrentinit/ConcurrentInitialize.java

```diff
@@ -0,0 +1,92 @@
+package aQute.lib.concurrentinit;
+
+/**
+ * Helper class to handle concurrent system where you need to initialize a
+ * value. The first one should create the value but the others should block
+ * until the value has been created. Since we do not want to hold a lock during
+ * the creation this is kind of tricky. This class uses a single monitor
+ * {@link #lock} that oeprates a state machine.
+ *
+ * @param <T>
+ */
+public abstract class ConcurrentInitialize<T> {
+	enum State {
+		/*
+		 * Initial state, the first one that detects the machine is in this
+		 * state must create the object
+		 */
+		INIT,
+		/*
+		 * The object is being created, block until errored or created
+		 */
+		CREATING,
+		/*
+		 * There is an object, just return it
+		 */
+		DONE,
+		/*
+		 * There was an error during creation, throw the error
+		 */
+		ERROR
+	};
+
+	private State		state	= State.INIT;
+	private T			value;
+	private Object		lock	= new Object();
+	private Thread		creatingThread;
+	private Exception	exception;
+
+	/**
+	 * Get the value or wait until it is created.
+	 */
+	public T get() throws Exception {
+		synchronized (lock) {
+			switch (state) {
+				case INIT :
+					state = State.CREATING;
+					creatingThread = Thread.currentThread();
+					break;
+
+				case CREATING : {
+					if (creatingThread == Thread.currentThread())
+						throw new IllegalStateException(
+								"Cycle:  ConcurrentInitialize's create returns to same instance");
+					do {
+						try {
+							lock.wait();
+						} catch (InterruptedException e) {
+							return null;
+						}
+					} while (state == State.CREATING);
+					return value;
+				}
+
+				case ERROR :
+					throw exception;
+
+				case DONE :
+					return value;
+			}
+		}
+		try {
+			value = create();
```

Comment:
I fixed it but in this case it is not a problem. Any other thread is blocked until the state progresses from CREATING. Since it is a private variable as far as I can see nobody can use the object before the subsequent synchronised block has exited?

Anyway, I moved the synchronized blocks into one method to make it simpler